go/pkg/cas/fscache.go

```diff
@@ -0,0 +1,85 @@
+package cas
+
+import (
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/golang/protobuf/proto"
+
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/cache"
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/digest"
+	repb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+)
+
+// FSCache caches information collected/computed from the file system.
+//
+// Currently does not take ModTime into account and assumes that file system
+// does not change. TODO(nodir): improve this.
+//
+// Zero value is a valid cache.
+type FSCache struct {
+	// c is the primary cache storage.
+	// Keys are of type fsCacheKey and values are of type fsCacheValue.
+	// See also makeKey.
+	c cache.SingleFlight
+}
+
+type fsCacheKey struct {
+	AbsPath       string
+	ExcludeRegexp string
+}
+
+type fsCacheValue struct {
+	dirEntry proto.Message // FileNode, DirectoryNode or SymlinkNode
+	digest   *repb.Digest  // may be nil if a symlink is dangling
+}
+
+// makeKey returns a key for FSCache.c.
+func (c *FSCache) makeKey(absPath string, mode os.FileMode, pathExclude *regexp.Regexp) fsCacheKey {
+	key := fsCacheKey{
+		AbsPath: absPath,
+	}
+
+	// Incorporate the pathExclude, unless it is a regular file.
+	// If it is a regular file, then there's no need to include pathExclude in the
+	// cache key, as we already know the regex does not match the file and the
+	// exclusion isn't propagated.
+	if pathExclude != nil && !mode.IsRegular() {
+		key.ExcludeRegexp = pathExclude.String()
+	}
+
+	return key
+}
+
+// Digest returns the digest computed for a file/dir at ps.Path.
+//
+// In the context of upload, ps.Exclude must match one of the PathSpecs passed
+// to Client.Upload earlier.
+//
+// If the digest is unknown, returns (nil, nil).
+// If the file is a danging symlink, then its digest is unknown.
+func (c *FSCache) Digest(ps *PathSpec) (*digest.Digest, error) {
+	absPath, err := filepath.Abs(ps.Path)
```

Comment:
I've sent out https://github.com/bazelbuild/remote-apis-sdks/pull/345 to require `PathSpec.Path` to be absolute, but for a different reason.

I am not sure I understood your reasoning though. Most package I know, including the lowest-level `os` package, accept relatives. Does your argument apply to them too?

Reply:
My argument does apply to them too - relative paths are tricky because we have clear definitions of "input/exec root" and a "working directory", but, AFAIK, no expectations for the running directory of the binary using the SDK.

Reply:
the path is absolute now.